### ISSUE TITLE
Speed issue with elpy

### ISSUE THREAD

**Issue by teymuri** [jorgenschaefer/elpy#1475]
# Summary
Is there an speed issue with elpy? It makes emacs 25 very slow, so that even moving the cursor around happens with some delay for me. There is no speed issue when i disable elpy. 

Thank you
Thank you for reporting to Elpy !

Here is a template to help you do it so your issue can be fixed as
soon as possible. Please replace this paragraph with a summary of
your issue and fill the next sections.

# Steps to reproduce

# My configuration
## Result of `(elpy-config)`
```
# Paste it here
```

## Elpy configuration in my init.el
```
# Paste it here
```

**Comment by ateymuri** [comment]
What part of the `elpy-config` do you want me to send?

**Comment by ateymuri** [comment]
Elpy Configuration

Virtualenv........: None
RPC Python........: 3.5.3 (/usr/bin/python3)
Interactive Python: python3 (/usr/bin/python3)
Emacs.............: 25.1.1
Elpy..............: 1.25.0
Jedi..............: 0.13.1
Rope..............: Not found (0.11.0 available)
Autopep8..........: Not found (1.4 available)
Yapf..............: Not found (0.24.0 available)
Black.............: Not found (18.9b0 available)
Syntax checker....: Not found (flake8)

You have not activated a virtual env. While Elpy supports this, it is
often a good idea to work inside a virtual env. You can use M-x
pyvenv-activate or M-x pyvenv-workon to activate a virtual env.

The directory ~/.local/bin/ is not in your PATH. As there is no active
virtualenv, installing Python packages locally will place executables
in that directory, so Emacs won't find them. If you are missing some
commands, do add this directory to your PATH -- and then do
`elpy-rpc-restart'.

The autopep8 package is not available. Commands using this will not
work.

[run] python3 -m pip install --user autopep8

The yapf package is not available. Commands using this will not work.

[run] python3 -m pip install --user yapf

The black package is not available. Commands using this will not work.

[run] python3 -m pip install --user black

The configured syntax checker could not be found. Elpy uses this
program to provide syntax checks of your programs, so you might want
to install one. Elpy by default uses flake8.

[run] python3 -m pip install --user flake8

Options

`Raised' text indicates buttons; type RET or click mouse-1 on a button
to invoke its action.  Invoke [+] to expand a group, and [-] to
collapse an expanded group.  Invoke the [Group], [Face], and [Option]
buttons below to edit that item in another window.

[+]-- Group Elpy
[+]-- Group Python
[+]-- Group Virtual Environments (Pyvenv)
[+]-- Group Completion (Company)
[+]-- Group Call Signatures (ElDoc)
[+]-- Group Inline Errors (Flymake)
[+]-- Group Snippets (YASnippet)
[+]-- Group Directory Grep (rgrep)
[+]-- Group Search as You Type (ido)
[+]-- Group Django extension
[+]-- Group Autodoc extension

**Comment by galaunay** [comment]
Is it happening if you set `(setq elpy-company-add-completion-from-shell nil)` ?

**Comment by stardiviner** [comment]
I still feel elpy is a little slow about 2 seconds to raise up completions.

Here is my Emacs profiler result:

- [X] memory report

```
- elpy-rpc--filter                                                                  10,746,130  30%
 - elpy-rpc--handle-json                                                            10,637,218  29%
  - elpy-promise-resolve                                                            10,637,218  29%
   - elpy-promise-reject                                                            10,280,318  28%
    - elpy-rpc--default-error-callback                                              10,278,233  28%
     - signal                                                                       10,277,209  28%
      - debug                                                                       10,277,209  28%
       - recursive-edit                                                              9,282,500  26%
        - command-execute                                                            3,178,700   8%
         - call-interactively                                                        3,178,700   8%
          - funcall-interactively                                                    3,178,668   8%
           - save-buffer                                                             3,178,668   8%
            - basic-save-buffer                                                      3,178,668   8%
             - read-file-name                                                        3,176,620   8%
              - apply                                                                3,176,620   8%
               - ad-Advice-read-file-name                                            3,176,620   8%
                - #<compiled 0x24b78f>                                               3,176,620   8%
                 - read-file-name-default                                            3,176,620   8%
                  - completing-read                                                  3,172,410   8%
                   - ivy-completing-read                                             3,172,410   8%
                    - ivy-read                                                       3,171,386   8%
                     - read-from-minibuffer                                          2,988,512   8%
                      - redisplay_internal (C function)                                 14,818   0%
                       - eval                                                            8,674   0%
                        - let*                                                           8,674   0%
                         - list                                                          8,530   0%
                          - *buffer-name                                                 8,482   0%
                           - propertize                                                  8,482   0%
                            + format                                                     8,482   0%
                          + *buffer-info                                                    48   0%
                         + concat                                                           96   0%
                         + propertize                                                       48   0%
                       + org-table-sticky-header--scroll-function                        6,144   0%
                      + #<compiled 0x2d60d51>                                           14,075   0%
                      + ivy--queue-exhibit                                              14,075   0%
                      + winner-save-unconditionally                                      4,136   0%
                        minibuf-eldef-setup-minibuffer                                   3,072   0%
                      + rfn-eshadow-update-overlay                                       3,072   0%
                      + command-execute                                                     16   0%
                     + ivy--reset-state                                                176,136   0%
                      replace-regexp-in-string                                           1,024   0%
                    abbreviate-file-name                                                 2,162   0%
                  + minibuffer-maybe-quote-filename                                      2,048   0%
             - run-hooks                                                                 2,048   0%
              - time-stamp                                                               2,048   0%
                 time-stamp-once                                                         1,024   0%
        - winner-save-old-configurations                                                 6,248   0%
           winner-remember                                                               6,248   0%
        - redisplay_internal (C function)                                                  972   0%
         - eval                                                                            972   0%
          - let*                                                                           972   0%
           - list                                                                          684   0%
            - *buffer-name                                                                 588   0%
             - propertize                                                                  588   0%
              - format                                                                     588   0%
               - s-truncate                                                                588   0%
                + if                                                                       588   0%
            + *buffer-info                                                                  96   0%
           + concat                                                                        192   0%
           + propertize                                                                     96   0%
       - pop-to-buffer                                                                 918,672   2%
        - display-buffer                                                               918,672   2%
         - display-buffer-below-selected                                               850,064   2%
          - window--try-to-split-window                                                730,061   2%
           - split-window-sensibly                                                     730,061   2%
            - split-window-below                                                       730,061   2%
             - split-window                                                            730,061   2%
              - visual-fill-column-split-window                                        730,061   2%
               - split-window                                                          730,061   2%
                - split-window-internal                                                  6,144   0%
                 - org-table-sticky-header--scroll-function                              6,144   0%
                  - org-table-sticky-header--fetch-header                                6,144   0%
                   - org-table-sticky-header-org-table-header-visible-p                  6,144   0%
                    - org-table-sticky-header--table-real-begin                          6,144   0%
                     - org-table-sticky-header--is-header-p                              5,120   0%
                        split-string                                                     1,024   0%
                     - org-table-begin                                                   1,024   0%
                      - cond                                                             1,024   0%
                       - save-excursion                                                  1,024   0%
                          and                                                            1,024   0%
          - window-in-direction                                                        119,238   0%
           - posn-at-point                                                               5,462   0%
            - eval                                                                       5,462   0%
             - let*                                                                      5,462   0%
              - list                                                                     5,430   0%
               - *buffer-name                                                            5,414   0%
                - propertize                                                             5,414   0%
                 - format                                                                5,414   0%
                  - s-truncate                                                           5,414   0%
                   - if                                                                  5,414   0%
                      file-relative-name                                                 5,414   0%
               - *buffer-info                                                               16   0%
                - concat                                                                    16   0%
                   let*                                                                     16   0%
              - concat                                                                      16   0%
               - *recursive-editing                                                         16   0%
                - if                                                                        16   0%
                   string-empty-p                                                           16   0%
              - propertize                                                                  16   0%
               - `                                                                          16   0%
                - list                                                                      16   0%
                 - list                                                                     16   0%
                  - list                                                                    16   0%
                   - +                                                                      16   0%
                      string-width                                                          16   0%
            window--display-buffer                                                         765   0%
           display-buffer-assq-regexp                                                   68,608   0%
       - debugger-setup-buffer                                                          71,136   0%
        - debugger-insert-backtrace                                                     68,136   0%
         - cl-prin1                                                                     68,136   0%
          - cl-print-object                                                             68,136   0%
           - apply                                                                      68,136   0%
            - #<compiled 0x28e9ee1>                                                     68,136   0%
             - apply                                                                    68,136   0%
              - #<compiled 0x2025731>                                                   68,136   0%
               - #<compiled 0x2adf88d>                                                  68,136   0%
                - apply                                                                 68,136   0%
                 - #<compiled 0x29970c1>                                                62,136   0%
                  - cl-print-object                                                     46,136   0%
                   - apply                                                              42,000   0%
                    - #<compiled 0x28e9f41>                                             42,000   0%
                     - apply                                                            42,000   0%
                      - #<compiled 0x2025731>                                           42,000   0%
                       - #<compiled 0x2da07d9>                                          42,000   0%
                        - apply                                                         42,000   0%
                           #<compiled 0x29970c1>                                        16,000   0%
                           #<compiled 0x2997099>                                        14,000   0%
                           #<compiled 0x1bd9cd1>                                        12,000   0%
                   + cl--generic-cache-miss                                              4,136   0%
                   #<compiled 0x2997099>                                                 6,000   0%
        - cl-prin1                                                                       3,000   0%
         - cl-print-object                                                               3,000   0%
          - apply                                                                        3,000   0%
           - #<compiled 0x28e9ee1>                                                       3,000   0%
            - apply                                                                      3,000   0%
             - #<compiled 0x2025731>                                                     3,000   0%
              - #<compiled 0x2adf88d>                                                    3,000   0%
               - apply                                                                   3,000   0%
                  #<compiled 0x29970c1>                                                  3,000   0%
       - #<compiled 0x2b77959>                                                           4,901   0%
        - quit-restore-window                                                            4,901   0%
         - window--delete                                                                4,901   0%
          - delete-window                                                                4,901   0%
           - run-window-configuration-change-hook                                        4,136   0%
            - window--adjust-process-windows                                             4,136   0%
             - window--process-window-list                                               4,136   0%
                walk-windows                                                             4,136   0%
      generate-new-buffer                                                                1,061   0%
    - #<compiled 0x2718d99>                                                              1,024   0%
     - kill-buffer                                                                       1,024   0%
        tramp-flush-file-function                                                        1,024   0%
   - #<compiled 0x168508d>                                                             310,364   0%
    - python-info-current-defun                                                        267,300   0%
     - python-nav-beginning-of-defun                                                   147,580   0%
      - python-nav--beginning-of-defun                                                 146,524   0%
       - python-nav-backward-block                                                     102,516   0%
        - python-nav-forward-block                                                     102,516   0%
         - python-nav-beginning-of-statement                                            36,448   0%
          - python-info-line-ends-backslash-p                                           17,608   0%
             syntax-ppss                                                                16,544   0%
            syntax-ppss                                                                  8,272   0%
         + python-syntax-context-type                                                    9,336   0%
       + python-info-looking-at-beginning-of-defun                                      30,072   0%
       + python-syntax-context-type                                                      5,200   0%
     - python-nav-end-of-defun                                                         108,400   0%
      - python-nav-end-of-statement                                                     87,032   0%
         syntax-ppss                                                                    41,360   0%
       - python-info-line-ends-backslash-p                                              14,536   0%
          syntax-ppss                                                                   12,408   0%
      - python-util-forward-comment                                                     14,536   0%
         syntax-ppss                                                                    12,408   0%
     - python-nav-beginning-of-statement                                                 7,304   0%
      - python-info-line-ends-backslash-p                                                4,136   0%
         syntax-ppss                                                                     4,136   0%
    - eldoc-message                                                                     30,720   0%
     - eldoc-minibuffer-message                                                         30,720   0%
        apply                                                                           30,720   0%
    - version<=                                                                         10,272   0%
       version-to-list                                                                  10,272   0%
   - #<compiled 0x2a2e439>                                                              42,400   0%
    - company-dabbrev-code                                                              29,056   0%
     - company-dabbrev--search                                                          29,056   0%
      - company-dabbrev--search-buffer                                                  23,840   0%
       - company-in-string-or-comment                                                    4,136   0%
          syntax-ppss                                                                    4,136   0%
    - elpy-company--cache-completions                                                   13,344   0%
     - mapcar                                                                           13,344   0%
      - #<compiled 0x2872489>                                                           13,344   0%
         company-grab-symbol                                                            12,288   0%
 - json-read                                                                           107,280   0%
  - json-read-object                                                                   107,280   0%
   - json-read                                                                          99,092   0%
      json-read-number                                                                  48,432   0%
      json-read-keyword                                                                 23,728   0%
    - json-read-array                                                                   16,376   0%
     - json-read                                                                        16,376   0%
        json-read-object                                                                16,376   0%
    - json-read-object                                                                  10,556   0%
     - json-read                                                                        10,556   0%
        json-read-number                                                                10,556   0%
```

- [X] CPU report

```
- timer-event-handler                                                                      996  24%
 - apply                                                                                   993  24%
  - company-idle-begin                                                                     732  18%
   - company-auto-begin                                                                    728  17%
    - company--perform                                                                     728  17%
     - company--begin-new                                                                  728  17%
      - company-calculate-candidates                                                       677  16%
       - company--fetch-candidates                                                         675  16%
        - sit-for                                                                          603  14%
         - read-event                                                                      546  13%
          - elpy-rpc--filter                                                               424  10%
           - elpy-rpc--handle-json                                                         424  10%
            - elpy-promise-resolve                                                         424  10%
             - #<compiled 0x29020b1>                                                       406  10%
              - company-dabbrev-code                                                       406  10%
               - company-dabbrev--search                                                   394   9%
                - company-dabbrev--search-buffer                                           394   9%
                 - company-in-string-or-comment                                            111   2%
                    syntax-ppss                                                            107   2%
             - #<compiled 0x168508d>                                                        18   0%
              - python-info-current-defun                                                   12   0%
               - python-nav-beginning-of-defun                                               8   0%
                - python-nav--beginning-of-defun                                             8   0%
                 - python-info-looking-at-beginning-of-defun                                 5   0%
                    syntax-ppss                                                              5   0%
                 - python-nav-backward-block                                                 3   0%
                  - python-nav-forward-block                                                 3   0%
                   - python-nav-beginning-of-statement                                       3   0%
                    - python-info-line-ends-backslash-p                                      3   0%
                       syntax-ppss                                                           3   0%
               - python-nav-beginning-of-statement                                           4   0%
                  syntax-ppss                                                                4   0%
              - eldoc-message                                                                6   0%
               - eldoc-minibuffer-message                                                    6   0%
                  apply                                                                      6   0%
          - timer-event-handler                                                            117   2%
           - apply                                                                         115   2%
            - #<compiled 0x280fc7>                                                          42   1%
             - eldoc-print-current-symbol-info                                              42   1%
              + eldoc-message                                                               24   0%
              - elpy-eldoc-documentation                                                    18   0%
               - elpy-rpc-get-calltip                                                       18   0%
                - elpy-rpc                                                                  18   0%
                 - elpy-rpc--call                                                           18   0%
                  - json-encode                                                             18   0%
                   - json-encode-list                                                       18   0%
                    - json-encode-alist                                                     18   0%
                     - mapcar                                                               13   0%
                      - #<compiled 0x15c0859>                                               13   0%
                         json-encode-key                                                     8   0%
                       json-join                                                             5   0%
        - #<compiled 0x2a4f9e9>                                                             58   1%
         - #<compiled 0x2a509e9>                                                            58   1%
          - elpy-rpc-get-completions                                                        56   1%
           - elpy-rpc                                                                       56   1%
            - elpy-rpc--call                                                                52   1%
             - json-encode                                                                  52   1%
              - json-encode-list                                                            52   1%
               - json-encode-alist                                                          52   1%
                - mapcar                                                                    52   1%
                 - #<compiled 0x15c0859>                                                    52   1%
                  - json-encode                                                             26   0%
                   - json-encode-list                                                       17   0%
                    - json-encode-array                                                     17   0%
                     - mapconcat                                                            17   0%
                      - json-encode                                                         17   0%
                       - json-encode-string                                                  8   0%
                          apply                                                              4   0%
                     json-encode-number                                                      6   0%
                     json-encode-string                                                      3   0%
                  - json-encode-key                                                         16   0%
                   - json-read-from-string                                                  16   0%
                      generate-new-buffer                                                   10   0%
                      json-read                                                              4   0%
                    - #<compiled 0x2fb8199>                                                  2   0%
                     - kill-buffer                                                           2   0%
                      - replace-buffer-in-windows                                            2   0%
                         unrecord-window-buffer                                              2   0%
        - company-call-backend-raw                                                          12   0%
         - apply                                                                            12   0%
          - company--multi-backend-adapter                                                  11   0%
           - company--multi-backend-adapter-candidates                                      11   0%
            - company-yasnippet                                                             11   0%
             - company-yasnippet--candidates                                                11   0%
              - company-yasnippet--completions-for-prefix                                    6   0%
               - cl-mapcan                                                                   6   0%
                - mapcan                                                                     6   0%
                 - #<compiled 0x1bb9789>                                                     6   0%
                  - maphash                                                                  6   0%
                     #<compiled 0x1bb9761>                                                   2   0%
              + yas--get-snippet-tables                                                      3   0%
          + company-files                                                                    1   0%
       + company-call-backend                                                                2   0%
      + company--multi-backend-adapter                                                      36   0%
      + company-call-backend                                                                15   0%
     company-post-command                                                                    4   0%
  - #<compiled 0x280fc7>                                                                   111   2%
   - eldoc-print-current-symbol-info                                                       111   2%
    - eldoc-message                                                                         59   1%
     - eldoc-minibuffer-message                                                             59   1%
        apply                                                                               59   1%
    - elpy-eldoc-documentation                                                              52   1%
     - elpy-rpc-get-calltip                                                                 52   1%
      - elpy-rpc                                                                            52   1%
       - elpy-rpc--call                                                                     46   1%
        - json-encode                                                                       46   1%
         - json-encode-list                                                                 46   1%
          - json-encode-alist                                                               46   1%
           - mapcar                                                                         37   0%
            - #<compiled 0x15c0859>                                                         37   0%
             - json-encode-key                                                              23   0%
              - json-read-from-string                                                       16   0%
               - #<compiled 0xe8f325>                                                       11   0%
                  kill-buffer                                                                9   0%
                 generate-new-buffer                                                         5   0%
                json-encode                                                                  2   0%
             + json-encode                                                                   9   0%
             json-join                                                                       6   0%
    jit-lock-context-fontify                                                                 3   0%
 - timer-activate-when-idle                                                                  3   0%
    timer--activate                                                                          3   0%
- elpy-rpc--filter                                                                         732  18%
 - elpy-rpc--handle-json                                                                   728  17%
  - elpy-promise-resolve                                                                   726  17%
   - elpy-promise-reject                                                                   337   8%
    - elpy-rpc--default-error-callback                                                     337   8%
     - signal                                                                              337   8%
      - debug                                                                              337   8%
       - recursive-edit                                                                    274   6%
        - command-execute                                                                  123   3%
         - call-interactively                                                              123   3%
          - funcall-interactively                                                          123   3%
           - save-buffer                                                                   123   3%
            - basic-save-buffer                                                            123   3%
             - read-file-name                                                              123   3%
              - apply                                                                      123   3%
               - ad-Advice-read-file-name                                                  123   3%
                - #<compiled 0x24b78f>                                                     123   3%
                 - read-file-name-default                                                  123   3%
                  - completing-read                                                        117   2%
                   - ivy-completing-read                                                   117   2%
                    - ivy-read                                                             117   2%
                     - read-from-minibuffer                                                 68   1%
                      - #<compiled 0x2d60d51>                                                9   0%
                       - ivy--minibuffer-setup                                               9   0%
                        - ivy--exhibit                                                       9   0%
                         - ivy--filter                                                       5   0%
                            ivy--re-filter                                                   5   0%
                         - ivy--insert-minibuffer                                            4   0%
                            ivy--resize-minibuffer-to-fit                                    4   0%
                      - ivy--queue-exhibit                                                   3   0%
                       - ivy--exhibit                                                        3   0%
                        - ivy--insert-minibuffer                                             3   0%
                         - ivy--resize-minibuffer-to-fit                                     3   0%
                          - window-resize                                                    3   0%
                           - window--resize-mini-window                                      3   0%
                            - window--pixel-to-total                                         3   0%
                             - window--pixel-to-total-1                                      3   0%
                              - window--pixel-to-total-1                                     3   0%
                               - window--pixel-to-total-1                                    3   0%
                                  window-combination-p                                       3   0%
                      - timer-event-handler                                                  3   0%
                         apply                                                               3   0%
                     - ivy--reset-state                                                     36   0%
                      - ivy--sorted-files                                                   33   0%
                       - sort                                                               17   0%
                          ivy-sort-file-function-default                                    17   0%
                       - all-completions                                                    16   0%
                        - read-file-name-internal                                           16   0%
                         - completion--some                                                 16   0%
                          - #<compiled 0x2de6ce9>                                           16   0%
                           - complete-with-action                                           16   0%
                            - completion--file-name-table                                   16   0%
                             - all-completions                                               9   0%
                                completion-file-name-table                                   9   0%
                             - completion--twq-all                                           7   0%
                              - mapcar                                                       7   0%
                               - #<compiled 0x2b594f1>                                       7   0%
                                - minibuffer-maybe-quote-filename                            7   0%
                                 - file-local-name                                           7   0%
                                    file-remote-p                                            7   0%
                      - ivy-thing-at-point                                                   3   0%
                       - ffap-file-at-point                                                  3   0%
                        - substitute-in-file-name                                            3   0%
                           substitute-env-in-file-name                                       3   0%
                    abbreviate-file-name                                                     6   0%
        - timer-event-handler                                                                5   0%
           apply                                                                             5   0%
       - pop-to-buffer                                                                      26   0%
        - display-buffer                                                                    26   0%
         - display-buffer-below-selected                                                    26   0%
            window-in-direction                                                             26   0%
       - debugger-setup-buffer                                                              23   0%
        - debugger-insert-backtrace                                                         23   0%
           symbol-file                                                                       7   0%
       - debugger-mode                                                                       3   0%
          kill-all-local-variables                                                           3   0%
   - #<compiled 0x168508d>                                                                 195   4%
    - eldoc-message                                                                        111   2%
     - eldoc-minibuffer-message                                                            111   2%
        apply                                                                              111   2%
    - python-info-current-defun                                                             84   2%
     - python-nav-beginning-of-defun                                                        52   1%
      - python-nav--beginning-of-defun                                                      52   1%
       - python-nav-backward-block                                                          31   0%
        - python-nav-forward-block                                                          31   0%
         - python-nav-beginning-of-statement                                                25   0%
          - python-info-line-ends-backslash-p                                                9   0%
             syntax-ppss                                                                     4   0%
            syntax-ppss                                                                      8   0%
            back-to-indentation                                                              1   0%
       - python-info-looking-at-beginning-of-defun                                          18   0%
          syntax-ppss                                                                       18   0%
     - python-nav-end-of-defun                                                              27   0%
      - python-util-forward-comment                                                         15   0%
         syntax-ppss                                                                         7   0%
      - python-nav-end-of-statement                                                          8   0%
         syntax-ppss                                                                         8   0%
     - python-nav-beginning-of-statement                                                     5   0%
      - python-info-line-ends-backslash-p                                                    5   0%
         syntax-ppss                                                                         5   0%
   - #<compiled 0x3180895>                                                                 194   4%
    - company-dabbrev-code                                                                 194   4%
     - company-dabbrev--search                                                             191   4%
      - company-dabbrev--search-buffer                                                     188   4%
       - company-in-string-or-comment                                                       61   1%
          syntax-ppss                                                                       61   1%
 - json-read                                                                                 1   0%
    json-read-object                                                                         1   0%
- ...                                                                                      359   8%
   Automatic GC                                                                            359   8%
- eldoc-pre-command-refresh-echo-area                                                      262   6%
 - eldoc-message                                                                           262   6%
  - eldoc-minibuffer-message                                                               262   6%
     apply                                                                                 262   6%
- redisplay_internal (C function)                                                          183   4%
 - jit-lock-function                                                                       139   3%
  - jit-lock-fontify-now                                                                   135   3%
   - jit-lock--run-functions                                                               125   3%
    - run-hook-wrapped                                                                     125   3%
     - #<compiled 0x1cec845>                                                               125   3%
      - font-lock-fontify-region                                                           122   3%
       - font-lock-default-fontify-region                                                  122   3%
        - font-lock-fontify-keywords-region                                                 76   1%
         - #<lambda 0x16a56abc4>                                                            12   0%
            hl-todo--search                                                                  6   0%
           #<compiled 0x2374e99>                                                             9   0%
        - font-lock-fontify-syntactically-region                                            28   0%
           syntax-ppss                                                                       9   0%
         - python-font-lock-syntactic-face-function                                          9   0%
          - python-info-docstring-p                                                          9   0%
           - python-nav-beginning-of-statement                                               6   0%
            - python-info-line-ends-backslash-p                                              3   0%
               syntax-ppss                                                                   3   0%
           + python-info-assignment-statement-p                                              3   0%
           looking-back                                                                      3   0%
 - eval                                                                                     26   0%
  + let*                                                                                    26   0%
 + #<compiled 0x44fd51>                                                                     15   0%
```

And my command `elpy-config` result:

```
Elpy Configuration

Virtualenv........: python3.7 (/home/stardiviner/.virtualenvs/python3.7)
RPC Python........: 3.7.0 (/home/stardiviner/.virtualenvs/python3.7/bin/python)
Interactive Python: python (/home/stardiviner/.virtualenvs/python3.7/bin/python)
Emacs.............: 26.1
Elpy..............: 1.25.0
Jedi..............: 0.13.1
Rope..............: Not found (0.11.0 available)
Autopep8..........: Not found (1.4.1 available)
Yapf..............: Not found (0.24.0 available)
Black.............: Not found (18.9b0 available)
Syntax checker....: flake8 (/home/stardiviner/.virtualenvs/python3.7/bin/flake8)

The autopep8 package is not available. Commands using this will not work.

[run] python -m pip install autopep8

The yapf package is not available. Commands using this will not work.

[run] python -m pip install yapf

The black package is not available. Commands using this will not work.

[run] python -m pip install black

Options

`Raised' text indicates buttons; type RET or click mouse-1 on a button to invoke its
action. Invoke [+] to expand a group, and [-] to collapse an expanded group. Invoke the
[Group], [Face], and [Option] buttons below to edit that item in another window.

[+]-- Group Elpy
[+]-- Group Python
[+]-- Group Virtual Environments (Pyvenv)
[+]-- Group Completion (Company)
[+]-- Group Call Signatures (ElDoc)
[+]-- Group Inline Errors (Flymake)
[+]-- Group Snippets (YASnippet)
[+]-- Group Directory Grep (rgrep)
[+]-- Group Search as You Type (ido)
[+]-- Group Django extension
[+]-- Group Autodoc extension
```

**Comment by galaunay** [comment]
Thanks for providing all of that.

> I still feel elpy is a little slow about 2 seconds to raise up completions.

Is Emacs hanging while waiting for completion ?

The call to `elpy-rpc--default-error-callback` make me think about errors in the RPC, do you have any message from Elpy in your `*Messages*` buffer ?

Apart from that, I can't find anything wrong.
Another possibility may be jedi being slow because... it is sometimes slow...

It would be interesting to have a comparison between Elpy and jedi "on its own" regarding the time they take to provide completionG.
You can benchmark jedi with something like that:
```Python
import jedi

filepath = 'test.py'
line = 7
column = 5
with open(filepath, 'r') as f:
    source = "".join(f.readlines())
script = jedi.Script(source=source, path=filepath, line=line, column=column)
candidates = script.completions()
print([cand.name for cand in candidates])
```

**Comment by stardiviner** [comment]
Emacs is not hanging when waiting for completion.

I benchmarked jedi with your python code.
But it does not give out benchmark timing data as result. It just give out candidates. So I modified it by adding `time` to check out completion time:

Here is my test and result:

#+begin_src python :tangle "/tmp/test.py"
#!/usr/bin/env python -v
# -*- coding: utf-8 -*-

import os
import cv2
import jedi
import json
import matplotlib.pyplot as plt

def testHello(name, word):
    """test hello."""
    print(name, word)
    pri

testHello("chris", "hello")

os.getenv("HOME")
#+end_src

#+begin_src python :results output
import jedi
import time

filepath = '/tmp/test.py'
# at line position "pri"
line = 13
column = 7
with open(filepath, 'r') as f:
    source = "".join(f.readlines())
script = jedi.Script(source=source, path=filepath, line=line, column=column)

start_time = time.time()

candidates = script.completions()
print([cand.name for cand in candidates])

end_time = time.time()

elismiated_time = end_time - start_time
print(elismiated_time)
#+end_src

#+RESULTS[<2018-10-19 17:06:35> e1908966fa5462bb5015d3800c33af539bb82a95]:
: ['print']
: 0.3862776756286621

This time, it feels faster now. Don't know why.

Then I run profiler again, found the `elpy-rpc--default-error-callback` is gone (not in stack). This is weird too. Anyway, it is fast now. Sorry for occupying your time.

**Comment by galaunay** [comment]
Well, I do like issues that disappear by themselves :).

Feel free to reopen if it happens again.

**Comment by vitorqb** [comment]
Just for those who may end up here with a similar problem (as I did), just wanted to register something that seemed to have worked for me. I was facing the same problem after updating to emacs 26.1 (and also updating many elisp packages). Company started to take ~2 seconds to show completions, which were almost instantaneous before.

As the problem did not happen with `emacs -q`, I started experimenting with partial changes on my setup. In the end what solved the bug was removing a trailing slash from the virtual environment path. That is, I changed `(pyvenv-activate "/path/to/venv/")` to `(pyvenv-activate "/path/to/venv")`.

I have no idea how or why this solved the bug, neither if it makes sense at all (elpy was able to detect the virtual environment anyway with the trailing slash, as shown in the `elpy-config` buffer). But I can swear that company started working perfectly again after removing the trailing slash. I repeated the test many times and putting the slash back there causes the problem once again.
